text/1025-branches.rst

```diff
@@ -0,0 +1,181 @@
+   having database states that correspond to the schema and avoiding
+   accidental mismatches in migrations.
+
+   If the ``from <oldbranch>`` part is omitted, the new branch is created
+   using the branch from the current connection.
+
+3) ``drop branch <oldbranch>``
+
+   Removes an existing branch from the instance.
+
+
+Project and CLI Integration
+---------------------------
+
+Currently the project configuration file ``edgedb.toml`` does not have to
+specify the default *database* it connects to. With this proposal specifying
+the branch is mandatory.
+
+A new project should ask the user for the branch name defaulting to "main".
+
+The ``database`` CLI commands should be deprecated and new ``branch`` commands
+introduced instead.
+
+``edgedb branch create`` command options:
+
+* It must have the ``--from <oldname>`` option. If this option is omitted the
+  current branch specified in the ``edgedb.toml`` file is used as the "from"
+  branch. After the branch is created, the ``edgedb.toml`` file is updated to
+  use the new branch as the connection default.
+
+* It must have the ``--empty`` option. If specified, this is equivalent to
+  running ``create empty branch <newname>`` DDL command.
+
+* It should have the ``--copy-data`` option. By default a new branch copies
+  schema only.
+
+The old ``drop`` subcommand works pretty much same as before.
+
+The old ``wipe`` subcommand may still be relevant for resetting a particular
+branch.
+
+There must be a new ``set`` sub-command that allows changing the default
+branch to connect to by updating the ``edgedb.toml`` file.
+
+
+Implementation
+--------------
+
+Most of the ``branch`` functionality is either existing ``database``
+functionality or can be implemented on top of that by performing migrations.
+
+The option to copy the data can be implemented in Postgres by using the
+``CREATE DATABASE newname WITH TEMPLATE = oldname`` which may be preferable
+to a dump/restore as it should be faster. The caveat is that unlike with a
+dump/restore the template database cannot have any other active connections.
+However, this may be reasonable for local development as the benefit is speed.
+
+
+Rebasing Branches
+-----------------
+
+We need to be able to rebase and merge database branches. This overlaps a lot
+with the scope of the ``migration`` commands.
+
+When rebasing one branch on top of another we can use introspection to compare
+the respective migration histories and find the point where they diverge.
+Afterwards we can try to apply a batch of new migration to the existing branch
+and if there are no issues perform a "fast-forward" rebase.
+
+In order to minimize the hassle of rebasing the git branches corresponding to
+database branches we need to give migration files names that are distinct in
+these branches so that git does not attempt to merge the file contents. This
+may be either directly the branch name appended to the numeric name or a
+(short) hash derived from the branch name.
+
```

Comment:
This is how I understand this process:

Main (VCS and database) branch contains migrations M1, M2 and M3.
They are written in files named 00001.edgeql, 00002.edgeql, 00003.edgeql, respectively.
These filenames do not contain branch name, as that name is equal to `main`.

Now we start a new (VCS and database) branch feat-a for feature A and
create migrations A1, A2 and A3,
in files feat-a-00004.edgeql, feat-a-00005.edgeql and feat-a-00006.edgeql.
These files might have been added in a single or in multiple VCS commits.

In parallel, we continue development on main branch with migrations M4 and M5, in files
00004.edgeql and 00005.edgeql.

Now we want to merge (rebase) feature A into main.
Normal VCS process would start on branch main and merge files changes from feature A.
Because migration files contain branch names, this will not merge the two 00004.edgeql files into one.

After this step, the file system would contain files 00001.edgeql, ..., 00005.edgeql,
as well as feat-a-00004.edgeql, feat-a-00005.edgeql and feat-a-00006.edgeql.

Because files feat-a-00004.edgeql and 00004.edgeql both reference M3 as their parent, we need
to rebase all of feat-a files onto M5.
For this, we'd introduce `edgedb branch rebase feat-a`, which would:
- read all migration files,
- identify all migrations from feat-a, by filenames,
- verify that all remaining migrations are linear and have been applied to the database,
- recreate feat-a migrations, one by one,
- write them to 00006.edgeql, 00007.edgeql and 00008.edgeql,
- delete original feat-a files.